lib/rust/ensogl/component/text/src/font.rs

```diff
@@ -0,0 +1,617 @@
+#[derivative(Clone(bound = ""))]
+pub struct FontTemplate<Family, Variations> {
+    rc: Rc<FontTemplateData<Family, Variations>>,
+}
+
+/// Internal representation of [`FontTemplate`].
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct FontTemplateData<Family, Variations> {
+    pub name:                   Name,
+    pub family:                 Family,
+    pub atlas:                  msdf::Texture,
+    pub cache:                  RefCell<HashMap<Variations, FontDataCache>>,
+    // FIXME: remove after MSDF-gen API will be updated to handle GlyphIds.
+    pub glyph_id_to_code_point: RefCell<HashMap<GlyphId, char>>,
+}
+
+/// A cache for common glyph properties, used to layout glyphs.
+#[derive(Debug, Default)]
+pub struct FontDataCache {
+    kerning: HashMap<(GlyphId, GlyphId), f32>,
+    glyphs:  HashMap<GlyphId, GlyphRenderInfo>,
+}
+
+impl<F, V> From<FontTemplateData<F, V>> for FontTemplate<F, V> {
+    fn from(t: FontTemplateData<F, V>) -> Self {
+        let rc = Rc::new(t);
+        Self { rc }
+    }
+}
+
+impl<F: Family<V>, V: Eq + Hash + Clone> FontTemplate<F, V> {
+    /// Constructor.
+    pub fn new(name: Name, family: impl Into<F>) -> Self {
+        let atlas = default();
+        let cache = default();
+        let family = family.into();
+        let glyph_id_to_code_point = default();
+        let data = FontTemplateData { name, family, atlas, cache, glyph_id_to_code_point };
+        Self { rc: Rc::new(data) }
+    }
+
+    /// Get the glyph id of the provided code point.
+    pub fn glyph_id_of_code_point(&self, variations: &V, code_point: char) -> Option<GlyphId> {
+        self.family
+            .with_borrowed_face(variations, |face| {
+                let id = face.ttf.as_face_ref().glyph_index(code_point);
+                if let Some(id) = id {
+                    self.glyph_id_to_code_point.borrow_mut().insert(id, code_point);
+                }
+                id
+            })
+            .flatten()
+    }
+
+    /// Get render info for one character, generating one if not found.
+    pub fn glyph_info(&self, variations: &V, glyph_id: GlyphId) -> Option<GlyphRenderInfo> {
+        let opt_render_info =
+            self.cache.borrow().get(variations).and_then(|t| t.glyphs.get(&glyph_id)).copied();
+        if opt_render_info.is_some() {
+            opt_render_info
+        } else {
+            self.family.update_msdfgen_variations(variations);
+            self.family.with_borrowed_face(variations, |face| {
+                let render_info = GlyphRenderInfo::load(&face.msdf, glyph_id, &self.atlas);
+                if !self.cache.borrow().contains_key(variations) {
+                    self.cache.borrow_mut().insert(variations.clone(), default());
+                }
+                self.cache
+                    .borrow_mut()
+                    .get_mut(variations)
+                    .unwrap()
+                    .glyphs
+                    .insert(glyph_id, render_info);
```

Comment:
1.Please "split" it to 2-3 lines.
2. I would borrow it once before the if. This way we avoid unnecessary borrow checking.